sql/parser: Adding descriptions to builtins

sploiselle opened the pull request:
Work on cockroachdb/docs#19

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12533)
<!-- Reviewable:end -->

Commit:
sql/parser: Adding descriptions to builtins

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=12533) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=12533) before we can accept your contribution.<br/>

a-robinson commented:
First commit to the code repo? 🎉🎉🎉

tamird commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/parser/builtins.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFab9pbMl0Y-R-t9u:-KZXFab9pbMl0Y-R-t9v:b-ggdla2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/builtins.go#L210)):*
> ```Go
> }
> 
> // TODO(dt/sean): Populate `Info` in all builtins.
> ```

should probably update this TODO

---

*[pkg/sql/parser/builtins.go, line 214 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXG7Hw7V3qod4KVjWr:-KZXG7Hw7V3qod4KVjWs:biccvis) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/builtins.go#L214)):*
> ```Go
> // Where we use a helper, eg `stringBuiltin1`, we'll need to plumb that though.
> 
> // TODO(dt/sean): Name all function arguments, switching `ArgTypes{a, b, ...}`
> ```

looks like you've addressed this TODO as well, to some extent?

---

*[pkg/sql/parser/generator_builtins.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFg5VV6jqudgPaUMr:-KZXFg5VV6jqudgPaUMs:b2qcd83) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/generator_builtins.go#L112)):*
> ```Go
> }
> 
> func makeGeneratorBuiltin(in ArgTypes, ret TTuple, g generatorFactory, info string) Builtin {
> ```

you should document, either here or on the structure itself (or both), the intended use of `info` and the tense that should be used. Concretely, the current state is that some of the Info strings you've added use a passive voice, and some use an active one. For instance "Not usable; ..." is a passive voice, but "Calculate the smallest integer greater than `val`." is an active, imperative voice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Oh, and this probably needs tests, if `Info` is actually use anywhere at run time.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-:-KZXGCYIgGqoGWpBH5Qo:bkw4frp)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/parser/builtins.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFab9pbMl0Y-R-t9u:-KZXUaLqCpas6PEEqVNl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/builtins.go#L210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should probably update this TODO
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 214 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXG7Hw7V3qod4KVjWr:-KZXUbQ--1cF9HnzTuAQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/builtins.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

looks like you've addressed this TODO as well, to some extent?
</blockquote></details>

Done.

---

*[pkg/sql/parser/generator_builtins.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFg5VV6jqudgPaUMr:-KZXV589qzW_fbNONalP:b-xbdomy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/generator_builtins.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you should document, either here or on the structure itself (or both), the intended use of `info` and the tense that should be used. Concretely, the current state is that some of the Info strings you've added use a passive voice, and some use an active one. For instance "Not usable; ..." is a passive voice, but "Calculate the smallest integer greater than `val`." is an active, imperative voice.
</blockquote></details>

Added to [`builtins.go`](https://github.com/cockroachdb/cockroach/pull/12533/files#diff-a78b90ae27ed7c44b4c7c4ceaa691cdbR122)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/builtins.go, line 124 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXWwAJtzyant7cZNTX:-KZXWwAJtzyant7cZNTY:bv5tb92) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/builtins.go#L124)):*
> ```Go
> 	// Info is a description of the function, which is surfaced on the CockroachDB
> 	// docs site on the "Functions and Operators" page. Descriptions should use
> 	// imperative syntax, but should focus more on ease of understanding so other
> ```

apparently the term is "imperative mood" https://en.wikipedia.org/wiki/Imperative_mood

---

*[pkg/sql/parser/generator_builtins.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFg5VV6jqudgPaUMr:-KZXWrX1kGULR-EKCPWD:b42iuwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/generator_builtins.go#L112)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Added to [`builtins.go`](https://github.com/cockroachdb/cockroach/pull/12533/files#diff-a78b90ae27ed7c44b4c7c4ceaa691cdbR122)
</blockquote></details>

Thanks. Still, can you help me understand why the descriptions are all second-person "Calculate foo and bar" rather than third-person "Calculate**s** foo and bar"?

---

*[pkg/sql/parser/generator_builtins.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXXRxBl0kxaxWg_ZMJ:-KZXXRxBl0kxaxWg_ZMK:bzov9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/generator_builtins.go#L105)):*
> ```Go
> var generators = map[string][]Builtin{
> 	"pg_catalog.generate_series": {
> 		makeGeneratorBuiltin(ArgTypes{TypeInt, TypeInt}, TTuple{TypeInt}, makeSeriesGenerator,
> ```

I think it's time to make these keyed literals https://golang.org/ref/spec#Composite_literals

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

dt commented:
Info (and Category()/category) aren't currently used at runtime, though they hopefully will be (and thus tests) if/when we add built-in help to cli.

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-:-KZX_mkL-Rm_d6OGZX8I:bk5921m)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/generator_builtins.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXXRxBl0kxaxWg_ZMJ:-KZXaCbBOlOvFf0kvfET:bulxzv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/generator_builtins.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think it's time to make these keyed literals https://golang.org/ref/spec#Composite_literals
</blockquote></details>

To clarify, you mean just de-helper these?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/generator_builtins.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXXRxBl0kxaxWg_ZMJ:-KZY3ReeHrfMr2Cu8Hx9:baego5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/generator_builtins.go#L105)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

To clarify, you mean just de-helper these?
</blockquote></details>

I guess so, or else just don't wrap these lines. all the braces on the preceding line contribute to the reading difficulty

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/generator_builtins.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFg5VV6jqudgPaUMr:-KZbBkb2VJXqULKHMWF-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/generator_builtins.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Thanks. Still, can you help me understand why the descriptions are all second-person "Calculate foo and bar" rather than third-person "Calculate**s** foo and bar"?
</blockquote></details>

Done.

---

*[pkg/sql/parser/generator_builtins.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXXRxBl0kxaxWg_ZMJ:-KZbBCHUuLTmSgt1z_zr:b-ip2v34) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/generator_builtins.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I guess so, or else just don't wrap these lines. all the braces on the preceding line contribute to the reading difficulty
</blockquote></details>

now one-arg-per-line, following the long-signature wrapping guide.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

dt commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 124 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXWwAJtzyant7cZNTX:-KZYDdynv495NPTATeMv:b33qgrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/builtins.go#L124)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

apparently the term is "imperative mood" https://en.wikipedia.org/wiki/Imperative_mood
</blockquote></details>

Changed verb tense to third-person.

---

*[pkg/sql/parser/generator_builtins.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXFg5VV6jqudgPaUMr:-KZb007ngRTy4A66JElV:b-g7hujk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1a0fb717d46b3560f611c48975a1d7c7c3a3dfd/pkg/sql/parser/generator_builtins.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Thanks. Still, can you help me understand why the descriptions are all second-person "Calculate foo and bar" rather than third-person "Calculate**s** foo and bar"?
</blockquote></details>

Second-person is what technical writing defaults to. In this case, think you're right that third person is a better choice, so changed the verb tense in the descriptions and updated the comment.

---

*[pkg/sql/parser/generator_builtins.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXXRxBl0kxaxWg_ZMJ:-KZYDfom_AZsY8YHLoIm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21bfde26f6f22faf53278dfe7c3e1012d5822929/pkg/sql/parser/generator_builtins.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I guess so, or else just don't wrap these lines. all the braces on the preceding line contribute to the reading difficulty
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/aggregate_builtins.go, line 178 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbFU-oyeF0jRsLZhvK:-KZbFU-oyeF0jRsLZhvL:beyg7jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/parser/aggregate_builtins.go#L178)):*
> ```Go
> 
> func makeAggBuiltins(f func() AggregateFunc, info string, types ...Type) []Builtin {
> 	ret := make([]Builtin, len(types))
> ```

hm, this usage (and all the usage above) suggests that `Info` should be a property of `[]Builtin` rather than `Builtin`.

ugh, doesn't apply to functions overloaded for strings and byte arrays :(

---

*[pkg/sql/parser/builtins.go, line 309 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbG6AkuXHsRGRUV5g4:-KZbG6AlxApDuTrhkbte:b-i4udjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/parser/builtins.go#L309)):*
> ```Go
> 				return NewDString(buf.String()), nil
> 			},
> 			Info: "Uses the first string as a separator between the concatenation of the " +
> ```

s/string/argument/g?

---

*[pkg/sql/testdata/pgoidtype, line 14 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbGbtvvfjH-hGT6y8k:-KZbGbtvvfjH-hGT6y8l:b-rwdo6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/testdata/pgoidtype#L14)):*
> ```
> SELECT 'upper'::REGPROC, 'upper'::REGPROCEDURE
> ----
> 1736923753  1736923753
> ```

how come this changed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/aggregate_builtins.go, line 178 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbFU-oyeF0jRsLZhvK:-KZbHW4Q5SFqOmd9AFO4:b-6ndeej) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/parser/aggregate_builtins.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, this usage (and all the usage above) suggests that `Info` should be a property of `[]Builtin` rather than `Builtin`.

ugh, doesn't apply to functions overloaded for strings and byte arrays :(
</blockquote></details>

yeah, it is definitely not always the case that all overloads have the same description.

there've been some discussions on how to group similar overloads during rendering though -- e.g. those that _do_ have the same doc string get folded together or something -- but we want to keep the raw representation able to capture different descriptions for each impl.

---

*[pkg/sql/testdata/pgoidtype, line 14 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbGbtvvfjH-hGT6y8k:-KZbHJhClTpt6p3jn7Lx:b-envz4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/testdata/pgoidtype#L14)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come this changed?
</blockquote></details>

because it includes the hash of input type, which changed from the legacy simple arg to the new named arg.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/parser/builtins.go, line 309 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbG6AkuXHsRGRUV5g4:-KZbIAN8FU4j1-Mw9CZd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e667f8e937fe9c1f85f10ac552967806426d8131/pkg/sql/parser/builtins.go#L309)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/string/argument/g?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/builtins.go, line 310 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbImp0U-_KrcI3BCqj:-KZbImp1xK22RJhSRjPn:blty6ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96c224e79e54279ca59d365543acc4acdbf5748/pkg/sql/parser/builtins.go#L310)):*
> ```Go
> 			},
> 			Info: "Uses the first argument as a separator between the concatenation of the " +
> 				"subsequent strings. <br/><br/>For example `concat_ws('!','wow','great')` " +
> ```

here too (which is what the `g` at the end of s/foo/bar/g typically indicates)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/parser/builtins.go, line 310 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZbImp0U-_KrcI3BCqj:-KZbMgKetWj9xiT6ynsd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96c224e79e54279ca59d365543acc4acdbf5748/pkg/sql/parser/builtins.go#L310)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

here too (which is what the `g` at the end of s/foo/bar/g typically indicates)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -1459,92 +1615,110 @@ var substringImpls = []Builtin{
 		return NewDFloat(DFloat(math.Pow(x, y))), nil
-	}, ""),
-	decimalBuiltin2(func(x, y *inf.Dec) (Datum, error) {
+	}, "Calculate `x`^`y`."),
```
For consistency with other descriptions: "Calculates"

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -1459,92 +1615,110 @@ var substringImpls = []Builtin{
 			y := int64(*args[1].(*DInt))
 			return NewDInt(DInt(intPow(x, y))), nil
 		},
+		Info: "Calculates `x`^`y`",
```
Missing period.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -1101,37 +1222,41 @@ var Builtins = map[string][]Builtin{
+		}, "Rounds `val` to the nearest integer."),
 		Builtin{
-			Types:      ArgTypes{TypeFloat, TypeInt},
+			Types:      NamedArgTypes{{"input", TypeFloat}, {"decimcal_accuracy", TypeInt}},
```
Should `decimcal_accuracy` be `decimal_accuracy`? If so, this also needs to be updated in the description.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -1101,37 +1222,41 @@ var Builtins = map[string][]Builtin{
 			fn: func(_ *EvalContext, args DTuple) (Datum, error) {
 				return round(float64(*args[0].(*DFloat)), int64(*args[1].(*DInt)))
 			},
+			Info: "Keeps `decimcal_accuracy` number of figures to the right of the zero position " +
```
Same as above.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"&#8226; **g**: Global matching (match each substring instead of only the " +
+				"first).<br/><br/>&#8226; **i**: Case-insensitive matching<br/><br/>&#8226; " +
+				"**m** or **n**: Newline-sensitive matching: `.` and negated brackets (`[^...]`)" +
+				" brackets do not match newline characters (preventing matches from crossing " +
```
Remove duplicate "brackets".

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"first).<br/><br/>&#8226; **i**: Case-insensitive matching<br/><br/>&#8226; " +
+				"**m** or **n**: Newline-sensitive matching: `.` and negated brackets (`[^...]`)" +
+				" brackets do not match newline characters (preventing matches from crossing " +
+				"newlines unless explicitly defined to); `^` and `$` match the space before and" +
```
Need space after final "and". Otherwise, in rendered version, you get "andafter".

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"**m** or **n**: Newline-sensitive matching: `.` and negated brackets (`[^...]`)" +
+				" brackets do not match newline characters (preventing matches from crossing " +
+				"newlines unless explicitly defined to); `^` and `$` match the space before and" +
+				"after newline characters respectively (so characters between newline characters" +
```
Same as above: Need space after final "characters".

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				" brackets do not match newline characters (preventing matches from crossing " +
+				"newlines unless explicitly defined to); `^` and `$` match the space before and" +
+				"after newline characters respectively (so characters between newline characters" +
+				"are treated as if they're on a separate line).<br/><br/>&#8226; **p**: Partial" +
```
Need space after "Partial". Otherwise, this gets rendered as "Partialnewline-sensitive".

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"newlines unless explicitly defined to); `^` and `$` match the space before and" +
+				"after newline characters respectively (so characters between newline characters" +
+				"are treated as if they're on a separate line).<br/><br/>&#8226; **p**: Partial" +
+				"newline-sensitive matching: `.` and negated brackets (`[^...]`) do not match" +
```
Need space after "match". Same reason as above.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"after newline characters respectively (so characters between newline characters" +
+				"are treated as if they're on a separate line).<br/><br/>&#8226; **p**: Partial" +
+				"newline-sensitive matching: `.` and negated brackets (`[^...]`) do not match" +
+				"newline characters (preventing matches from crossing newlines unless explicitly" +
```
Need space after "explicitly". Same reason as above.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"`val`.<br/><br/>&#8226; **s**: Newline-insensitive matching" +
+				"*(default)*.<br/><br/>&#8226; **w**: Inverse partial newline-sensitive matching:" +
+				"`.` and negated brackets (`[^...]`) *do* match newline characters, but  `^` and" +
+				"`$` match the space before and after newline characters respectively (so" +
```
Need space after "so". Same reason as above.

jseldess commented on pkg/sql/parser/builtins.go:
```diff
@@ -580,16 +642,35 @@ var Builtins = map[string][]Builtin{
+				"*(default)*.<br/><br/>&#8226; **w**: Inverse partial newline-sensitive matching:" +
+				"`.` and negated brackets (`[^...]`) *do* match newline characters, but  `^` and" +
+				"`$` match the space before and after newline characters respectively (so" +
+				"characters between newline characters are treated as if they're on a separate" +
```
Need space after final "separate". Same reason as above. 

Please check for other similar instances. I probably missed some.

jseldess reviewed (approved):
LGTM, with some minor copyedits.

sploiselle commented:
Review status: 5 of 6 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/builtins.go, line 651 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-651:-KZbsQktu0uhKFSuT4UP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L651)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Remove duplicate "brackets".
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 652 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-652:-KZbt3viVUErmbAJUKZY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L652)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after final "and". Otherwise, in rendered version, you get "andafter".
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 653 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-653:-KZbu16RFSlGtKOEcJm8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L653)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Same as above: Need space after final "characters".
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 654 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-654:-KZbu1WwWNd4CjrGztMZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L654)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after "Partial". Otherwise, this gets rendered as "Partialnewline-sensitive".
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 655 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-655:-KZbu1rPHMmBtbeFkA76:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L655)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after "match". Same reason as above.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 656 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-656:-KZbu2Tu8it01lGPJseP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L656)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after "explicitly". Same reason as above.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 661 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-661:-KZbu3QeAlcx1S28UwOA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L661)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after "so". Same reason as above. 
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 662 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-662:-KZbu3u3Vu32jiVeGGxb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L662)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Need space after final "separate". Same reason as above. 

Please check for other similar instances. I probably missed some.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 772 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-772:-KZbuZJ6pCXqcbY_qX13:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L772)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

In our docs for `SERIAL`, which uses the `unique_rowid` function to generate default values, we mention that the default value "is the combination of the insert timestamp and the ID of the node executing the insert. This combination is guaranteed to be globally unique." Might be worth mentioning here, too.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 1236 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-1236:-KZbujGkyrWLjbUb425u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L1236)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Should `decimcal_accuracy` be `decimal_accuracy`? If so, this also needs to be updated in the description.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 1241 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-1241:-KZbujYkIvf_yP9cllZr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L1241)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 1700 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-1700:-KZburlAQWBqgDRRGV8A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L1700)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

For consistency with other descriptions: "Calculates"
</blockquote></details>

Done.

---

*[pkg/sql/parser/builtins.go, line 1717 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533#-KZXCusHba35dQKT52OU-r5-1717:-KZbuwNUXYgy1eXTMi6C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec5ac64c55a4457264354a2403c51a7f78d8b6b7/pkg/sql/parser/builtins.go#L1717)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Missing period.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12533)*
<!-- Sent from Reviewable.io -->
